### PR TITLE
feat: Action for DB restore

### PR DESCRIPTION
**Pull request by saurabhjambhule**

### CODE REVIEW

File: src/config.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            )
+    return data
+
+
+def load_target(data):
+    utils.validate_key(data, "Target", "Target")
+    utils.validate_key(data["Target"], "DBIdentifier", "DBIdentifier")
+
+    # if data["Target"].get("MasterUserPassword") is None:
+    #     data["Target"]["MasterUserPassword"] = generate_password()
```
Comment (ndibek) [inline src/config.py:44]: Do we need this commented out block?

File: src/config.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            if tf_outputs.get(result.group(2)) is None:
+                LOGGER.error("%s does not exists in TF state" % result.group(2))
+            value = tf_outputs[result.group(2)]
+        elif result.group(1) == "ssm":
+            value = ssm.get_parameter(assume_role, result.group(2))
+        elif result.group(1) == "env":
+            if os.environ.get(result.group(2)) is None:
+                LOGGER.error(
+                    "%s environment variable does not exists" % result.group(2)
+                )
```
Comment (ndibek) [inline src/config.py:108]: I would avoid nested if blocks of combination of for/if being nested.  Makes code hard to read, troubleshoot and maintain

File: src/main.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import tf
+import utils
+
+LOGGER = utils.setup_custom_logger("root")
+
+
+def main(command_line=None):
+    parser = argparse.ArgumentParser(description="Restore RDS snapshot")
+    parser.add_argument("-c", "--config", required=True)
+    parser.add_argument("-t", "--tfstate")
```
Comment (ndibek) [inline src/main.py:14]: Love the usage of the ArgumentParser lib here.  Should `--tfstate` be required too?

File: src/main.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    # target_data = {
+    #         'DBIdentifier': data["Target"]["DBIdentifier"],
+    #         'VpcSecurityGroupIds': data["Target"]["VpcSecurityGroupIds"],
+    #         'DBSubnetGroupName': data["Target"]["DBSubnetGroupName"],
+    #         'CopyTagsToSnapshot': True,
+    #         'DBInstanceClass': data["Target"]["DBInstanceClass"],
+    #         'PubliclyAccessible': False,
+    #         'Tags': [{'Key': 'dbrestore', 'Value': 'enable'}],
+    #         'SnapshotArn': 'arn:aws:rds:us-east-1:683046045338:snapshot:dash-staging-282751'
+    #     }
```
Comment (ndibek) [inline src/main.py:68]: We should remove this unused commented out code

**Review by ndibek — CHANGES_REQUESTED** [review]
Few comments on the python code.  I would recommend adding unit tests to this that verify and document expected behavior from this code

File: src/main.py (hunk 3 above)
Comment (saurabhjambhule) [reply src/main.py:14]: Nope, its optional

File: src/main.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    data = config.load_config(args.config)
+    config.validate(data)
+
+    if data["Target"].get("AssumeRole") is not None:
+        LOGGER.info("Assuming target AWS role")
+        data["Target"]["AssumeRole"] = utils.assume_aws_role(
+            data["Target"]["AssumeRole"]
+        )
+    else:
+        data["Target"]["AssumeRole"] = None
```
Comment (ndibek) [inline src/main.py:26]: Couple of comments here:  
1. Generally the arguments we get from the outside - call to the program and parsed into data hash array should be read only.  We can create other variables to inject values into an pass to the code bellow.  So instead of setting `data["Target"]["AssumeRole"] = utils.assume_aws_role(data["Target"]["AssumeRole"])`  we could simply set `target_credentials = utils.assume_aws_role(data["Target"]["AssumeRole"])
2. Second point I would like to make is that the `utils.assume_aws_role` method is actually returning credentials.  Why not call it `get_aws_credentials`.  
3. Even more the method `get_aws_credentials` could then decide whether to create and return aws credentials if the role passed is not None, otherwise return None.  That way eliminating the whole if/else block above to simply:
`target_credentials = utils.get_aws_credentials(data["Target"]["AssumeRole"])`

To me it makes this code a lot simpler and readable when we want to maintain and reuse it.

File: src/main.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    if data["Source"]["Share"].get("AssumeRole") is not None:
+        LOGGER.info("Assuming source AWS role")
+        data["Source"]["AssumeRole"] = utils.assume_aws_role(
+            data["Source"]["Share"]["AssumeRole"]
+        )
+    else:
+        data["Source"]["AssumeRole"] = None
```
Comment (ndibek) [inline src/main.py:34]: Same as previous, just call the variable `share_credentials` and get value from `utils.get_aws_credentials`

File: src/main.py (hunk 5 above)
Comment (ndibek) [reply src/main.py:26]: One more comment.  If you move the:
```python
    source = data["Source"]
    target = data["Target"]
```
above this code block then both this and the `data["Source"]["Share"]["AssumeRole"]` block bellow can be simply:
```python
    source = data["Source"]
    target = data["Target"]

    target_credentials = utils.get_aws_credentials(target["AssumeRole"])
    source_share_credentials = utils.get_aws_credentials(source["Share"])
```
Then anywhere you use `data["Target"]["AssumeRole"] ` you can replace it with `target_credentials` and same for  `data["Source"]["Share"]["AssumeRole"]` => ` source_share_credentials`

File: src/main.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    if args.tfstate is not None:
+        tf_outputs = tf.get_outputs(data["Target"]["AssumeRole"], args.tfstate)
+    data = config.replace_placeholder(data, tf_outputs, data["Target"]["AssumeRole"])
+
+    source = data["Source"]
+    target = data["Target"]
+    target_client = rds.init_client(target["AssumeRole"])
+
+    target_exists = rds.does_target_exists(
+        target_client, target["DBIdentifier"], data["ClusterMode"]
```
Comment (ndibek) [inline src/main.py:46]: Love this - very meaningful simple check, where name state what the code checks!

**Comment by ndibek** [comment]
In addition to specific code examples I would like to make a more generic comment:
As we as an organization move more towards an "Infrastructure as code" paradigm, where besides the code that runs our applications we own code that builds/monitors and scales our infrastructure (this being a service in part provided by Heroku before but now for the most part our code),  we need to make sure that Infrastructure code follows the same quality, simplicity and maintainability we expect in the app code. 
To me that means not just the simple refactoring cleanup suggestion that allow developers that have not written code to support it in the future.
But how do we make sure that the code we write works for edge cases, does not break at runtime and if we are building reusable components how do we document the calls to these.
Simply put that us a purpose of automated testing.  We have several Python classes, components here that we need to make sure properly run for any expected input.  I would definitely add a set of pytests that verify and confimr expected behavior of the code.

File: tests/test_config.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            "DBSubnetGroupName": "${tf:db_staging_private_subnet}",
+            "CopyTagsToSnapshot": True,
+            "DBInstanceClass": "db.t3.medium",
+            "PubliclyAccessible": False,
+            "Tags": [{"Key": "owner", "Value": "snapshot_restore"}],
+        },
+    }
+
+    status = is_valid(data)
+    assert status == True
```
Comment (ndibek) [inline tests/test_config.py:36]: There is a nice concept of text fixture that can be kind of a reusable data shared by multiple tests, so for example
```suggestion
@pytest.fixture
def data():
    return {
        "ClusterMode": False,
        "DeleteExistingTarget": True,
        "Source": {
            "Share": {
                "AssumeRole": "arn:aws:iam::3211251512:role/db_restore_share_role",
                "TargetAccount": "11223334544",
                "SourceKmsKey": "alias/db_restore",
                "TargetKmsKey": "alias/aws/rds",
            },
            "DBIdentifier": "dash-staging",
        },
        "Target": {
            "AssumeRole": "arn:aws:iam::11223334544:role/db_restore_role",
            "DBIdentifier": "dash-staging",
            "VpcSecurityGroupIds": "${tf:db_global_security_group}",
            "DBSubnetGroupName": "${tf:db_staging_private_subnet}",
            "CopyTagsToSnapshot": True,
            "DBInstanceClass": "db.t3.medium",
            "PubliclyAccessible": False,
            "Tags": [{"Key": "owner", "Value": "snapshot_restore"}],
        },
    }

def test_is_valid():
    assert is_valid(data) == True
```
While the advantage is not obvious on this first test when you look at the next tests that take `data` without some parts, then it becomes obvious that the test could be bit more readable, as we specify what is missing and how that impacts `is_valid`, like for example `test_is_valid_missing_required_properties` could be:
```python
def test_is_valid_missing_required_properties():
    invalid_data = data.copy().pop("ClusterMode)
    assert is_valid(invalid_data) == False
```
Note how we get valid object from the fixture and then create a copy of it without the "ClusterMode" value.
You can use this pattern in every validation test that checks for ay of the missing keys

**Review by ndibek — CHANGES_REQUESTED** [review]
We have tests!
We are getting closer,  I do have few more cleanup comments as well as comments on patterns we could use in tests to improve readability there too (keep in mind that the tests are generally what documents the code behavior)

One large comment -  and perhaps we could do this as a separate ticket, but I would definitely add running tests as a step in our build for this component.  Kind of like having a release gate.

File: src/config.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    except jsonschema.exceptions.ValidationError as err:
+        print(err)
+        return False
+    return True
+
+
+def load_config(filename):
+    if not is_config_exists(filename):
+        LOGGER.error("Provided configuration file is not available")
+        exit(1)
```
Comment (ndibek) [inline src/config.py:33]: We should not have program exit nested deep into the code.  If `is_config_exists` returns false, `load_config` could simply return `None` and `is_valid` can return false if there is no schema.

File: tests/test_config.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
+import pytest
+from src.config import *
+
```
[the remark this answers is not in the document]
Comment (saurabhjambhule) [reply tests/test_config.py:4]: Removed system exit from load_config. All tests are passing, added pytest GHA pipeline.

File: tests/test_tf.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    @pytest.fixture
+    @mock_s3
+    def conn(self):
+        return boto3.client("s3", region_name="us-east-1")
+
+    @mock_s3
+    def test_get_output_where_state_file_doesnt_exists(self, conn):
+        state_file = "g2dev-tf-state/staging/terraform.tfstate"
+        conn.create_bucket(Bucket="g2dev-tf-state")
+        assert get_outputs(ASSUME_ROLE_DISABLED, state_file) == None
```
Comment (ndibek) [inline tests/test_tf.py:18]: I would also add a test where a file exists on the s3 where the `get_outputs` has to parse the file.  You can fake existing file on s3 with moto and just a call to put_object (it has a param Body, and you just need stringified json)

File: src/main.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    target = data["Target"]
+    target_credentials = assume_aws_role(target["AssumeRole"], "target")
+    source_share_credentials = assume_aws_role(source["Share"]["AssumeRole"], "source")
+
+    tf_outputs = get_outputs(target_credentials, args.tfstate)
+    if tf_outputs is None:
+        LOGGER.error("TF state file does not exists")
+        exit(1)
+
+    source = replace_placeholder(source, tf_outputs, target_credentials)
```
Comment (ndibek) [inline src/main.py:39]: should the last argument here be source_share_credentials?

File: tests/test_config.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    def test_fetch_from_env_key_exists(self, tfdate):
+        key = "DB_USERNAME"
+        value = "random123"
+        os.environ[key] = value
+        assert fetch_from_env(key) == value
+
+    def test_fetch_from_env_key_doesnt_exists(self, tfdate):
+        key = "MASTER_DB_USERNAME"
+        value = "random123"
+        assert fetch_from_env(key) == None
```
Comment (ndibek) [inline tests/test_config.py:100]: `fetch_from_tfstate`, `fetch_from_env_key` and `fetch_from_ssm` are methods that seem to be used only by `replace_placeholder`, which is the only method called by the outside code.  Therefore I would treat these other 3 methods as private and not test them directly.
You can mark them as private by prefixing the name with underscore.
But you do not throw away the test and setup for these.
Just include the `test_fetch_from_tfstate` setup in a version of a test that calls `replace_placeholder` expecting it to call `fetch_from_tfstate` and so on.  That way we get coverage on all lines of `replace_placeholder` function and we get tests that are testing both `replace_placeholder` and the internal `fetch_...` functions.

**Review by ndibek — APPROVED** [review]
Practically ready for prime time.  Couple of more test suggestions, but that is it from me.  Seems like some of this code will be possible to reuse later too on other actions etc.

File: tests/test_config.py (hunk 13 above)
Comment (saurabhjambhule) [reply tests/test_config.py:100]: Got it, fetch_from_* is also tested using replace_placeholder.

File: src/main.py (hunk 12 above)
Comment (saurabhjambhule) [reply src/main.py:39]: The config, if is in SSM will be stored in the target account as here the DB will get created from the snapshot.
The source_share_credentials is optional and only needed when the source DB is located in a different AWS account.

File: tests/test_config.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+import pytest
+from src.config import *
+
+ASSUME_ROLE_DISABLED = None
+
+
+class TestConfig:
+    def test_is_config_exists_when_file_present(self):
+        assert is_config_exists("schema/config.json") == True
```
[the remark this answers is not in the document]
Comment (saurabhjambhule) [reply tests/test_config.py:11]: The test were failing with this when ran from root, so added method to get schema folder path

File: tests/test_config.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        assert is_invalid(invalid_data) is not None
+
+    def test_is_valid_missing_required_property_tags(self, data):
+        invalid_data = data.copy().pop("ClusterMode")
+        assert is_invalid(invalid_data) is not None
+
+    def test_is_valid_missing_optional_property_share(self, data):
+        valid_data = data.copy()
+        del valid_data["Source"]["Share"]
+        assert is_invalid(data) is False
```
Comment (ndibek) [inline tests/test_config.py:73]: You probably wanted to run the test on modified `valid_data` and not on the original `data`.

**Review by ndibek — APPROVED** [review]
LGTM!

File: src/config.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+from src.ssm import get_parameter
+
+LOGGER = logging.getLogger("root")
+
+
+def schema_dir():
+    return os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/schema/"
+
+
+def is_invalid(data):
```
Comment (etceterum) [inline src/config.py:18]: It is easier for me to think in terms of positive conditions, so I'd do a `def is_valid(data)` instead but maybe that's just me :) Feel free to ignore

File: src/config.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def is_sharing_enabled(data):
+    if data.get("Share") is None:
+        return False
+    return data["Share"]
+
+
+def _fetch_from_tfstate(key, tf_outputs):
+    if tf_outputs.get(key) is None:
```
Comment (etceterum) [inline src/config.py:52]: it'd make sense for me to call `tf_outputs.get(key)` and capture the value for return when it's not `None` instead of calling `tf_outputs[key]` later

File: src/config.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+        LOGGER.error("%s does not exists in TF state" % key)
+        return None
+    return tf_outputs[key]
+
+
+def _fetch_from_env(key):
+    if os.environ.get(key) is None:
```
Comment (etceterum) [inline src/config.py:59]: same as before

File: src/config.py [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+    schema = load_config(schema_dir() + "config.json")
+    try:
+        jsonschema.validate(data, schema)
+    except jsonschema.exceptions.ValidationError as err:
+        return err
+    return False
+
+
+def load_config(filename):
+    if not is_config_exists(filename):
```
Comment (etceterum) [inline src/config.py:31]: `does_config_exist`? :)

**Review by etceterum — COMMENTED** [review]
Left a few minor comments, feel free to ignore
